src/libsyntax/ptr.rs

```diff
@@ -63,45 +64,37 @@ impl<T: 'static> P<T> {
     }
 
     /// Produce a new `P<T>` from `self` without reallocating.
-    pub fn map<F>(mut self, f: F) -> P<T> where
+    pub fn map<F>(self, f: F) -> P<T> where
         F: FnOnce(T) -> T,
     {
-        let p: *mut T = &mut *self.ptr;
-
-        // Leak self in case of panic.
-        // FIXME(eddyb) Use some sort of "free guard" that
-        // only deallocates, without dropping the pointee,
-        // in case the call the `f` below ends in a panic.
-        mem::forget(self);
+        // If a panic occurs in `f`, leak the inner T.
+        let mut manually_drop_box = Box::into_manually_drop(self.ptr);
 
         unsafe {
-            ptr::write(p, f(ptr::read(p)));
-
-            // Recreate self from the raw pointer.
-            P { ptr: Box::from_raw(p) }
+            let x = f(ManuallyDrop::take(&mut manually_drop_box));
```

Comment:
Can you use `into_inner` instead of `take`?

Reply:
I don't think so, that would reallocate (or at least consume) the Box which defeats the point of P.

Reply:
It would move out of the box and then later move back into it, why would it reallocate? [This code does not reallocate either](https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=a69fa345f41e00d1428d83261e1e52fe).

Reply:
Oh wow. `Box` turns out to be even more magical than I realized. In this playground example shows that not only we can move a value out of `Box` without a `DerefMove` trait, but the compiler tracks such moves like for local variables and plain struct fields. Attempting to use the box before its contents are reassigned (if `*x = …` is removed) results in a compile time ``error[E0382]: use of moved value: `x` ``. But the box is still there and can be used again after its contents are reassigned.

And this is panic-safe too. In the assembly for https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=d4907435865f18159268fc36a8546016 we see that `foo` calls `box_free` but not `drop_t`.

So I think @RalfJung’s suggestion is the way to go.

Reply:
Bonus possibly-unexpected detail:

https://github.com/rust-lang/rust/blob/f22dca0a1bef4141e75326caacc3cd59f3d5be8e/src/liballoc/boxed.rs#L290-L295

Reply:
Wow, I was aware of `Box` special-casing, but I do not recall being able to reinitialize the `Box` and reuse it!

@nikomatsakis @pnkfelix Do you happen to know when that might've happened?

Reply:
I believe this probably happened as part of the NLL transition. We decided to stop hobbling the treatment of `Box` during that discussion.